### PR TITLE
add postrm script to deb and rpm packages

### LINKED ISSUE #3784: [upgrade] Deb package leaves stale symlink to elastic-agent
- .deb install
- Debian 11
- upgrade from 8.8.2 -> 8.10.4

After upgrade symlink to elastic-agent is wrong

```
/var/lib/dpkg/info# ls -l /usr/share/elastic-agent/bin/elastic-agent
lrwxrwxrwx 1 root root 62 Nov 2 12:46 /usr/share/elastic-agent/bin/elastic-agent -> /var/lib/elastic-agent/data/elastic-agent-cdc5ba/elastic-agent
```

symlink should point to:

```
/var/lib/elastic-agent/data/elastic-agent-a92ca3/elastic-agent 
```

looks like `elastic-agent.postinst` is to blame.

### LINKED ISSUE #4267: [uninstall] deb package leaves files behind even with --purge
If elastic-agent package is uninstalled with:

```
systemctl stop elastic-agent && apt-get -y -q purge elastic-agent
```

files are still left on the file system including:

```
/etc/elastic-agent
/etc/elastic-agent/elastic-agent.yml.2024-02-14T15-59-46.9261.bak
/etc/elastic-agent/fleet.enc
/etc/elastic-agent/fleet.enc.lock
/etc/elastic-agent/vault
/etc/elastic-agent/vault/ce5cae1525529517af95d07e1bf467e4c2971a9e80178b37178eade8754c5121
/etc/elastic-agent/vault/.lock
/etc/elastic-agent/vault/.seed
/usr/share/elastic-agent
/usr/share/elastic-agent/bin
/usr/share/elastic-agent/bin/elastic-agent
/var/lib/elastic-agent
/var/lib/elastic-agent/data
/var/lib/elastic-agent/data/agent.lock
/var/lib/elastic-agent/data/elastic-agent-ca4d78
/var/lib/elastic-agent/data/elastic-agent-ca4d78/logs
/var/lib/elastic-agent/data/elastic-agent-ca4d78/logs/elastic-agent-20240214-1.ndjson
/var/lib/elastic-agent/data/elastic-agent-ca4d78/logs/elastic-agent-20240214.ndjson
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/beat
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/beat/metrics-monitoring
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/beat/metrics-monitoring/meta.json
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/filestream-monitoring
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/filestream-monitoring/meta.json
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/filestream-monitoring/registry
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/filestream-monitoring/registry/filebeat
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/filestream-monitoring/registry/filebeat/log.json
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/filestream-monitoring/registry/filebeat/meta.json
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/http
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/http/metrics-monitoring
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/http/metrics-monitoring/meta.json
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/log-default
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/log-default/meta.json
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/log-default/registry
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/log-default/registry/filebeat
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/log-default/registry/filebeat/log.json
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/log-default/registry/filebeat/meta.json
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/system
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/system/metrics-default
/var/lib/elastic-agent/data/elastic-agent-ca4d78/run/system/metrics-default/meta.json
/var/lib/elastic-agent/data/elastic-agent-ca4d78/state.enc
/var/lib/elastic-agent/data/tmp
/var/log/elastic-agent
/var/log/elastic-agent/elastic-agent-20240214-1.ndjson
/var/log/elastic-agent/elastic-agent-20240214.ndjson
```

The use of `purge` should remove the files, if `uninstall` had been used it would be acceptable for the configuration files to remain.

### PR DESCRIPTION
**Pull request by leehinman**
## What does this PR do?

Adds a postrm script to the deb and rpm packages.  Script does the following.

On `rpm -e` and `apt-get remove` does the following:

- stops `elastic-agent` if it is running
- deletes symlink `/usr/share/elastic-agent/bin/elastic-agent` that post install script makes
- does a systemd daemon-reload, this is necessary since the systemd unit file will be removed.

on `apt-get purge` does the following:

- recursively deletes the following directories

  + `/var/lib/elastic-agent`
  + `/var/log/elastic-agent`
  + `/etc/elastic-agent`

## Why is it important?

1) leaving `elastic-agent` running after uninstall is unexpected, and policy updates may fail because necessary resources on disk may be missing

2) leaving the symlink has lead to updates failing

3) not doing `systemctl daemon-reload` after removing the elastic-agent unit files means systemd may still think it can control elastic-agent event though it doesn't have a unit file

4) in the case of removing the directories during an `apt-get purge`, this is necessary to completely remove all traces of elastic-agent from the disk

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

## Author's Checklist

- [ ]

## How to test this PR locally

1. Grab state of filesystem, something like `find / \( -path /proc -prune -o -path /home -prune -o -path /run -prune \) -o -print | sort > before.txt`
2. Install elastic-agent via deb or rpm (`rpm -i` or `dpkg -i`)
3. enroll elastic-agent
4. uninstall via deb or rpm (`rpm -e elastic-agent` or `apt-get remove elastic-agent`)
5. make sure elastic-agent process is stopped
6. grab state of filesystem now and diff with original
7. Deb only `apt-get purge elastic-agent`
8. grab state of filesystem now and diff with original

## Related issues

- Closes #3784
- Closes #4267

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: changelog/fragments/1708732907-add-postrm-script-to-deb-and-rpm-packages.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# - breaking-change: a change to previously-documented behavior
+# - deprecation: functionality that is being removed in a later release
+# - bug-fix: fixes a problem in a previous version
+# - enhancement: extends functionality but does not break or fix existing behavior
+# - feature: new functionality
+# - known-issue: problems that we are aware of in a given version
+# - security: impacts on the security of a product or a user’s deployment.
+# - upgrade: important information for someone upgrading from a prior version
+# - other: does not fit into any of the other categories
+kind: feature
```
Comment (blakerouse) [inline changelog/fragments/1708732907-add-postrm-script-to-deb-and-rpm-packages.yaml:11]:
```suggestion
kind: bug-fix
```

This really is a bug fix, this should have always done this.

File: changelog/fragments/1708732907-add-postrm-script-to-deb-and-rpm-packages.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+# Change summary; a 80ish characters long description of the change.
+summary: add postrm script to deb and rpm packages
```
Comment (blakerouse) [inline changelog/fragments/1708732907-add-postrm-script-to-deb-and-rpm-packages.yaml:14]:
```suggestion
summary: Fix issue uninstalling deb/rpm so it completely stops the service properly.
```

These are user facing so best to explain what it means to the user.

**Review by blakerouse — APPROVED** [review]
Looks good, thanks for working on this.

**Comment by leehinman** [comment]
> @leehinman should we backport this to 8.13 to have it fix on the way?

good catch, I added the label.

**Pull request merged**